docs/investigations/simplify-api-usage/simplify-api-usage.md

```diff
@@ -0,0 +1,141 @@
+
+|                    | Is `applicationID` provided? | Should pass? |
+|--------------------|:---------------:|--------------|
+| User               | YES             | YES          |
+| User               | NO              | NO           |
+| Application        | YES             | YES          |
+| Application        | NO              | YES          |
+| Runtime            | YES             | NO           |
+| Runtime            | NO              | NO           |
+| Integration system | YES             | YES          |
+| Integration system | NO              | NO           |
+
+There are 8 test cases in just one mutation, having in mind we would change around 20 mutations, it would be around 160 tests. 
+
+* API becomes less readable due to optional `ID` parameter (the user could be confused what happens if we don't provide one)
+* domains dependant on another mechanism which will fetch `consumer ID`
+
+### 2. Separated mutations where `consumer ID` input parameter is not needed
+There could be another mutation which doesn't accept `applicationID`/`runtimeID` as a parameter, so we would have doubled mutations. Take this for example:
+
+`addAPI(applicationID: ID, in: APIDefinitionInput! @validate): APIDefinition! @hasScopes(path: "graphql.mutation.addAPI")`
+
+and 
+
+`addAPIForApplication(in: APIDefinitionInput! @validate): APIDefinition! @hasScopes(path: "graphql.mutation.addAPI")`
+
+**Work that has to be done**
+* add new mutations to schema
+* implement the new mutations (for example for addAPIForApplication mutation we could reuse addAPI implementation)
+* test the new mutations (unit and integration)
+
+**Pros**
+* separated mutation for a special use case
+
+**Cons**
+* API becomes a lot bigger - around 20 new mutations
+* the same situation with tests as described in first solution
+
+### 3. The Viewer pattern
+The alternative to solutions presented above is the Viewer pattern.
+It is a GraphQL query that looks like:
+```graphql
+viewer: Viewer!
+type Viewer {
+  id: ID!
+  type: ViewerType!
+}
+```
+Usage:
+```graphql
+viewer {
+   id 
+   type // Application/Runtime/Integration System
+}
+
+```
+
+> The viewer field represents the currently logged-in user; its subfields expose data that are contextual to the user.
+
+The consumer ID is present in request context, so it can be retrieved from there.
+
+**Work that has to be done**
+* prepare and implement the Viewer query
+* test the query
+
+**Pros**
+* small amount of work
+* API doesn't get much bigger(only one query)
+
+**Cons**
+* consumer has to make an additional call to get its ID for other GraphQL operations 
+
+Sources:
+
+https://medium.com/workflowgen/graphql-schema-design-the-viewer-field-aeabfacffe72
+https://codeahoy.com/2019/10/13/graphql-practical-tutorial/
+
+The Viewer pattern is used in e.g. Facebook and GitHub API:
+
+https://developer.github.com/v4/query
+https://github.com/graphql/graphql-js/issues/571
+
```

Comment:
We should also describe, that even we have to perform 2 queries instead of one, this is not such a big issue at all.
In the case of the integration system, this endpoint won't be used, because it has to provide ID of the application explicitly.
In the case of the Application, if there is a library for performing GraphQL query, the overhead of performing another query is not too big in my opinion.
In the case of UI or user who use API directly, ID cannot be "guessed"

Reply:
I've added a paragraph describing that case